Adds RTC driver for Infineon XMC4xxx devices

talih0 opened the pull request:
Adds support for settings/getting RTC time and using alarm/update feature.
    
The alarm option needs all fields to be set due to a hardware limitation.
    
RTC shares the same interrupt with the watchdog. Thus shared
interrupts must be enabled when WDT and RTC both need to trigger the ISR.

Devicetree for boards xmc45_relax_kit and xmc47_relax_kit are also updated to include RTC.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/17433307/ec8261a8-87a5-402d-941f-99fa376ab1ca)

talih0 committed:
tests: drivers: rtc: Fix typos

Compare tm_nsec instead of tm_yday and fix error message.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented:
The compliance failure is a false positive. Though INTERUPT is misspelled it is referring to a function from Infineon hal.

bjarki-andreasen commented on drivers/rtc/rtc_xmc4xxx.c:
```diff
@@ -0,0 +1,266 @@
+{
+	const struct tm stdtime = { .tm_mday = 1, .tm_year = 1900};
+
+	if (!XMC_RTC_IsRunning()) {
+		if (!XMC_SCU_HIB_IsHibernateDomainEnabled()) {
+			XMC_SCU_HIB_EnableHibernateDomain();
+		}
+
+		XMC_RTC_SetPrescaler(RTC_XMC4XXX_DEFAULT_PRESCALER);
+		XMC_RTC_SetTimeStdFormat(&stdtime);
+
+		XMC_RTC_Start();
+	}
```
The RTC should not be started and set to a random value at init, time is set manually using the rtc_set_time api :)

bjarki-andreasen commented:
Please run the alarm test suite using this PR #78739 it allows you to set the alarm time mask using Kconfig to match the rtc :)

talih0 commented:
> Please run the alarm test suite using this PR #73789 it allows you to set the alarm time mask using Kconfig to match the rtc :)

thanks, I rebased after the commit was merged:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17433307/33de5f72-e95b-4078-a8a0-92a67b38e5f7)

bjarki-andreasen reviewed (changes requested):
Could you split the PR into
1. add device driver
2. add rtc to boards
3. add boards to RTC test suite

It keeps bisectability high and makes reviews a bit easier :)

talih0 committed:
drivers: rtc: Add RTC driver for Infineon XMC4xxx devices

Adds support for settings/getting RTC time and using alarm/update feature.
The alarm option needs all fields to be set due to a hardware limitation.

RTC shares the same interrupt with the watchdog. Thus shared
interrupts must be enabled when WDT and RTC both need to trigger the ISR.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
boards: infineon: Add RTC to xmc4(75)_relax_kit boards

Adds RTC node to xmc4(75)_relax_kits.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
tests: rtc_api: Add conf files for xmc4(75)_relax_kit boards

Adds conf files for xmc4(75)_relax_kits.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented on drivers/rtc/rtc_xmc4xxx.c:
```diff
@@ -0,0 +1,280 @@
+	struct rtc_xmc4xxx_data *dev_data = dev->data;
+
+	if (XMC_RTC_IsRunning()) {
+		dev_data->is_time_set = true;
```
yep, thanks.

bjarki-andreasen reviewed (dismissed):
Nice work!

bjarki-andreasen commented:
Please ping me after feature freeze if I forget about this PR, we need to ping around a bit to get it merged with a CI failure ;)

talih0 commented:
@bjarki-andreasen, I removed the function name that was causing the compliance problem.
